--- configure	2010-02-06 15:43:19 +0300
+++ configure	2011-01-21 19:27:34 +0300
@@ -5313,6 +5313,20 @@
   *-*-mingw*)
     gcc_cflags_optlist="$gcc_cflags_optlist nocygwin"
     gcc_cflags_nocygwin="-mno-cygwin"
+    if test "x$build" != "x" ; then
+      new_cclist=
+      for compiler in $cclist; do
+        new_cclist="$new_cclist $build-$compiler"
+      done
+      cclist="$new_cclist $cclist"
+      if test "x$cclist_64" != "x" ; then
+        new_cclist_64=
+        for compiler in $cclist_64; do
+          new_cclist_64="$new_cclist_64 $build-$compiler"
+        done
+        cclist_64="$new_cclist_64 $cclist_64"
+      fi
+    fi
     ;;
 esac
 
--- configure.in	2010-02-06 15:43:13 +0300
+++ configure.in	2011-01-21 19:27:47 +0300
@@ -1573,6 +1573,20 @@
   *-*-mingw*)
     gcc_cflags_optlist="$gcc_cflags_optlist nocygwin"
     gcc_cflags_nocygwin="-mno-cygwin"
+    if test "x$build" != "x" ; then
+      new_cclist=
+      for compiler in $cclist; do
+        new_cclist="$new_cclist $build-$compiler"
+      done
+      cclist="$new_cclist $cclist"
+      if test "x$cclist_64" != "x" ; then
+        new_cclist_64=
+        for compiler in $cclist_64; do
+          new_cclist_64="$new_cclist_64 $build-$compiler"
+        done
+        cclist_64="$new_cclist_64 $cclist_64"
+      fi
+    fi
     ;;
 esac
 
