Index: gcc/config/i386/t-mingw32
===================================================================
--- gcc/config/i386/t-mingw32   (revision 142523)
+++ gcc/config/i386/t-mingw32   (working copy)
@@ -1,5 +1,14 @@
 # Match SYSTEM_INCLUDE_DIR
 NATIVE_SYSTEM_HEADER_DIR = /mingw/include

+MULTILIB_OPTIONS = m64/m32
+MULTILIB_DIRNAMES = 64 32
+MULTILIB_OSDIRNAMES = ../lib64 ../lib
+#if TARGET_64BIT_DEFAULT
+#define MULTILIB_DEFAULTS { "m64" }
+#else
+#define MULTILIB_DEFAULTS { "m32" }
+#endif
+
 # MinGW-specific parts of LIB_SPEC
 SHLIB_LC = -lmingw32 -lmingwex -lmoldname -lmsvcrt -luser32 -lkernel32 -ladvapi32 -lshell32
Index: gcc/config/i386/mingw32.h
===================================================================
--- gcc/config/i386/mingw32.h   (revision 142523)
+++ gcc/config/i386/mingw32.h   (working copy)
@@ -50,10 +50,24 @@
 /* Override the standard choice of /usr/include as the default prefix
    to try when searching for header files.  */
 #undef STANDARD_INCLUDE_DIR
+
 #if TARGET_64BIT_DEFAULT
-#define STANDARD_INCLUDE_DIR "/mingw/include64"
+#define SPEC_32 "m32"
+#define SPEC_64 "!m32"
+
+#undef ASM_SPEC
+#define ASM_SPEC "%{v:-V} %{n} %{T} %{Ym,*} %{Yd,*} \
+ %{Wa,*:%*} %{m32:--32} %{m64:--64}"
+
 #else
+#define SPEC_32 "!m64"
+#define SPEC_64 "m64"
+#endif
+
+#if TARGET_64BIT_DEFAULT
 #define STANDARD_INCLUDE_DIR "/mingw/include"
+#else
+#define STANDARD_INCLUDE_DIR "/mingw/include"
 #endif
 #undef STANDARD_INCLUDE_COMPONENT
 #define STANDARD_INCLUDE_COMPONENT "MINGW"
@@ -80,7 +94,19 @@
 #define SUBTARGET_EXTRA_SPECS                                          \
   { "shared_libgcc_undefs", SHARED_LIBGCC_UNDEFS_SPEC }

-#define LINK_SPEC "%{mwindows:--subsystem windows} \
+#if TARGET_64BIT_DEFAULT
+#define SUB_LINK_SPEC "%{" SPEC_64 ":-m i386pep} %{" SPEC_32 ":-m i386pe}"
+#else
+#define SUB_LINK_SPEC ""
+#endif
+
+#define LINK_SPEC SUB_LINK_SPEC "%{mwindows:--subsystem windows} \
   %{mconsole:--subsystem console} \
   %{shared: %{mdll: %eshared and mdll are not compatible}} \
   %{shared: --shared} %{mdll:--dll} \